File: 2D Arrays (Matrix)/spiral_traverse.java
/*
	Write a function that takes in an n x m two-dimensional array (that can be square-shaped when n == m) and
	returns a one-dimensional array of all the array's elements in spiral order.

	Spiral order starts at the top left corner of the two-dimensional array, goes to the right, and proceeds
	in a spiral pattern all the way until every element has been visited.

	Explanation:

	The SpiralTraverse function takes a 2D integer array array and returns a 1D integer slice that contains the
	elements of array traversed in a spiral order, starting from the top-left corner and moving clockwise.

	The function first initializes an empty slice result to hold the elements of the spiral traversal.
	If the input array is empty, the function immediately returns the empty result.

	Next, the function initializes variables startRow, endRow, startCol, and endCol to keep track of the
	boundaries of the matrix. These variables will be updated as the function traverses the matrix.

	The function then enters a loop that traverses the matrix in a spiral order. The loop continues
	as long as startRow <= endRow and startCol <= endCol, which means that there are still elements
	in the matrix to be traversed.

	The first step in the loop is to traverse the top row of the matrix from left to right, and append
	each element to the result slice. The next step is to traverse the rightmost column of the matrix from top to bottom,
	and append each element to the result slice. If there is more than one row in the matrix, the function then traverses
	the bottom row of the matrix from right to left, and appends each element to the result slice. If there is only one row left,
	the loop is broken to avoid duplicating the elements. Finally, if there is more than one column in the matrix,
	the function traverses the left

	O(n) time | O(n) space - where n is the total number of elements in the array
*/
import java.util.*;
public class SpiralTraverse {

    public static void main(String[] args) {
        int[][] array = { {1, 2, 3},
                {12, 13, 4},
                {11, 14, 5},
                {10, 15, 6},
                {9, 8, 7} };
        System.out.println(spiralFillUsingLoop(array));

//        spiralFillUsingRecursion(array, 0, array.length - 1, 0, array[0].length - 1, result);
    }
    public static List<Integer> spiralFillUsingLoop(int[][] array) {
        // O(n) time | O(n) space;
        if(array.length == 0) return new ArrayList<>();

        var result = new ArrayList<Integer>();
        var startRow = 0;
        var endRow = array.length - 1;
        var startCol = 0;
        var endCol = array[0].length - 1;

        while(startRow <= endRow && startCol <= endCol) {

            for(int col = startCol; col <= endCol; col++)
                result.add(array[startRow][col]);

            for(int row = startRow + 1; row <= endRow; row++)
                result.add(array[row][endCol]);

            for(int col = endCol - 1; col >= startCol; col--) {
                /**
                 * Handle the edge case when there's a single row
                 * in the middle of the matrix, In this case. we don't
                 * want to double-count the values in this row, which
                 * we've already counted in the first for loop above.
                 * Test case for this edge case :
                 *                              [
                 *                                 [1, 2, 3, 4],
                 *                                 [10, 11, 12, 5],
                 *                                 [9, 8, 7, 6]
                 *                               ]
                 *
                 *
                 */
                if(startRow == endRow) break;
                result.add(array[endRow][col]);
            }

            for(int row = endRow - 1; row > startRow; row--) {
                /**
                 * Handle the edge case when there's a single column
                 * in the middle of the matrix, In this case. we don't
                 * want to double-count the values in this row, which
                 * we've already counted in the first for loop above.
                 * Test case for this edge case :
                 *                              [
                 *                                 [1, 2, 3],
                 *                                 [12, 13, 4],
                 *                                 [11, 14, 5],
                 *                                 [10, 15, 6],
                 *                               ]
                 *
                 */
                if(startCol == endCol) break;
                result.add(array[row][startCol]);
            }

            startRow++;
            endRow--;
            startCol++;
            endCol--;
        }

        return result;
    }
    public static void spiralFillUsingRecursion(
            int[][] array,
            int startRow,
            int endRow,
            int startCol,
            int endCol,
            ArrayList<Integer> result) {
        // O(n) time | O(n) space;
        if(startRow > endRow || startCol > endCol) return;

        for(int col = startCol; col <= endCol; col++)
            result.add(array[startRow][col]);
        for(int row = startRow + 1; row <= endRow; row++)
            result.add(array[row][endCol]);
        for(int col = endCol - 1; col >= startCol; col--) {
            if(startRow == endRow) break;
            result.add(array[endRow][col]);
        }
        for(int row = endRow - 1; row > startRow; row--) {
            if(startCol == endCol) break;
            result.add(array[row][startCol]);
        }
        spiralFillUsingRecursion(array, startRow + 1, endRow -1, startCol + 1, endCol - 1, result);
    }
}

File: 2D Arrays (Matrix)/spiral_traverse.py
##### LEETCODE 54. Spiral Matrix ####

##### INPUT OUTPUT #####

# Given an m x n matrix, return all elements of the matrix in spiral order.
# Input: matrix = [[1,2,3],[4,5,6],[7,8,9]]
# Output: [1,2,3,6,9,8,7,4,5]
# Input: matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
# Output: [1,2,3,4,8,12,11,10,9,5,6,7]
# Write python code along with explanation and comments use meaningful variables

##### EXPLAINATION #####

# The given problem statement is to return all elements of the matrix in spiral order. We can start by defining the four boundaries of the matrix - top, bottom, left and right. We can then traverse the matrix in a spiral order by following these four steps:

# Traverse from left to right along the top boundary
# Traverse from top to bottom along the right boundary
# Traverse from right to left along the bottom boundary
# Traverse from bottom to top along the left boundary
# After each traversal, we need to update the corresponding boundary and change the direction of traversal.

# We can implement this algorithm using a while loop that runs as long as the top boundary is less than or equal to the bottom boundary and the left boundary is less than or equal to the right boundary. Within the while loop, we can use an if-else ladder to check the current direction of traversal and perform the corresponding traversal along the boundary.

# Finally, we can return the result list containing all the spiral order elements of the matrix.


#### DRY RUN: ####

# Initially, we have the matrix:
# 1 2 3
# 4 5 6
# 7 8 9

# We initialize the variables top, bottom, left, and right to 0, 2, 0, and 2 respectively. We also set the direction variable to 0.
# Now, we enter the while loop since top<=bottom and left<=right.
# In the first iteration of the while loop, direction=0 means we need to traverse from left to right along the top boundary.
# We iterate the for loop from left=0 to right=2 and append the elements 1, 2, 3 to the result list. After this, we increment top by 1 to mark that the top boundary is now done.
# The result list now contains [1, 2, 3].
# In the second iteration of the while loop, direction=1 means we need to traverse from top to bottom along the right boundary.
# We iterate the for loop from top=1 to bottom=2 and append the elements 6, 9 to the result list. After this, we decrement right by 1 to mark that the right boundary is now done.
# The result list now contains [1, 2, 3, 6, 9].
# In the third iteration of the while loop, direction=2 means we need to traverse from right to left along the bottom boundary.
# We iterate the for loop from right=1 to left=0 and append the elements 8, 7 to the result list. After this, we decrement bottom by 1 to mark that the bottom boundary is now done.
# The result list now contains [1, 2, 3, 6, 9, 8, 7].
# In the fourth iteration of the while loop, direction=3 means we need to traverse from bottom to top along the left boundary.
# We iterate the for loop from bottom=1 to top=1 and append the element 4 and 5 to the result list. After this, we increment left by 1 to mark that the left boundary is now done.
# The result list now contains [1, 2, 3, 6, 9, 8, 7, 4, 5].
# Now, we have completed one full spiral traversal of the matrix. The direction variable is updated as (0+1)%4=1 which sets it to 1 for the next iteration.
# We continue with the while loop since top<=bottom and left<=right. In the second iteration, the process continues in the same way as described above until all elements of the matrix are visited.
# Finally, the function returns the result list containing all the spiral order elements of the matrix.

def spiralOrder(matrix):
    # Initialize variables to keep track of indices and boundaries
    top, bottom = 0, len(matrix) - 1
    left, right = 0, len(matrix[0]) - 1
    direction = 0  # 0 = right, 1 = down, 2 = left, 3 = up

    # Initialize an empty result list to store the spiral order elements
    result = []

    while top <= bottom and left <= right:
        if direction == 0:
            # Traverse right
            for i in range(left, right+1):
                result.append(matrix[top][i])
            top += 1
        elif direction == 1:
            # Traverse down
            for i in range(top, bottom+1):
                result.append(matrix[i][right])
            right -= 1
        elif direction == 2:
            # Traverse left
            for i in range(right, left-1, -1):
                result.append(matrix[bottom][i])
            bottom -= 1
        else:
            # Traverse up
            for i in range(bottom, top-1, -1):
                result.append(matrix[i][left])
            left += 1

        # Update direction after completing one full traversal
        direction = (direction + 1) % 4

    return result


matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
print(spiralOrder(matrix))
